### PR TITLE
Registry integration tests: Restrict

### LINKED ISSUE #897: Registry integration tests: Restrict
After the development of wazuh/wazuh#5576, new integration tests are going to be needed.
Only one file will be needed in this test:

### test_registry_restrict
| Test | Action |Expected result |
|---|---|---|
|test_restrict|Monitor a key using restrict option|Only events that match the restrict are generated

### PR DESCRIPTION
**Pull request by antoniomanuelfr**
Hello team,

This PR aims to add a new test for the windows registry integration tests. 

Closes #897 

# Tests logic
 ### First case
- Monitor two keys: One with `arch="64bit` and another with `arch="both"` and using a restrict for values.
- It creates values that match the regex and other values that don't match the regex and check that all the events are generated as expected.
### Second case
- Monitor two keys: One with `arch="64bit` and another with `arch="both"` and using a restrict for keys.
- It creates keys that match the regex and other keys that don't match the regex and check that all the events are generated as expected.
# Tests checks

- [X] Proven that tests **pass** when they have to pass
- [X] Proven that tests **fail** when they have to fail
- [ ] Tested and passed in Jenkins. Build URL: `<YOUR_JENKINS_BUILD_URL>`

## Windows

- When the expected behavior occurs:

```
python -m pytest -vvx test_restrict\
=========== test session starts =========== 
platform win32 -- Python 3.7.3, pytest-6.0.2, py-1.9.0, pluggy-0.13.0 -
[  2%] test_registry_restrict.py PASSED
[  4%] test_registry_restrict.py::test_restrict_value PASSED
[  8%] test_registry_restrict.py::test_restrict_value PASSED
[ 12%] test_registry_restrict.py::test_restrict_value PASSED
[ 16%] test_registry_restrict.py::test_restrict_value PASSED
[ 20%] test_registry_restrict.py::test_restrict_value PASSED
[ 25%] test_registry_restrict.py::test_restrict_key SKIPPED
[ 29%] test_registry_restrict.py::test_restrict_key SKIPPED
[ 33%] test_registry_restrict.py::test_restrict_key SKIPPED
[ 37%] test_registry_restrict.py::test_restrict_key SKIPPED
[ 41%] test_registry_restrict.py::test_restrict_key SKIPPED
[ 45%] test_registry_restrict.py::test_restrict_key SKIPPED
[ 50%] test_registry_restrict.py::test_restrict_value SKIPPED
[ 54%] test_registry_restrict.py::test_restrict_value SKIPPED
[ 58%] test_registry_restrict.py::test_restrict_value SKIPPED
[ 62%] test_registry_restrict.py::test_restrict_value SKIPPED
[ 66%] test_registry_restrict.py::test_restrict_value SKIPPED
[ 70%] test_registry_restrict.py::test_restrict_value SKIPPED
[ 75%] test_registry_restrict.py::test_restrict_key PASSED
[ 79%] test_registry_restrict.py::test_restrict_key PASSED
[ 83%] test_registry_restrict.py::test_restrict_key PASSED
[ 87%] test_registry_restrict.py::test_restrict_key PASSED
[ 91%] test_registry_restrict.py::test_restrict_key PASSED
[ 95%] test_registry_restrict.py::test_restrict_key PASSED
[100%]

===== 12 passed, 12 skipped in 25.62s =====

```

Best regards.

### CODE REVIEW

File: tests/integration/test_fim/test_registry/test_restrict/test_registry_restrict.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    key_h = create_registry(registry_parser[key], subkey, arch)
+    # Create values
+    modify_registry_value(key_h, value_name, REG_SZ, "added")
+    # Go ahead in time to let syscheck perform a new scan
+    check_time_travel(get_configuration['metadata']['fim_mode'] == 'scheduled', monitor=wazuh_log_monitor)
+
+    event = wazuh_log_monitor.start(timeout=global_parameters.default_timeout,
+                                callback=callback_detect_event, accum_results= 2 if triggers_event else 1).result()
+    if triggers_event:
+
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_registry/test_restrict/test_registry_restrict.py:89]: Please, remove this blank line.
```suggestion
```

File: tests/integration/test_fim/test_registry/test_restrict/test_registry_restrict.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    (key, sub_key_2, KEY_WOW64_64KEY, valid_value_name, True, {'value_restrict'}),
+    (key, sub_key_1, KEY_WOW64_64KEY, no_valid_value_name, False, {'value_restrict'}),
+    (key, sub_key_2, KEY_WOW64_32KEY, no_valid_value_name, False, {'value_restrict'}),
+    (key, sub_key_2, KEY_WOW64_64KEY, no_valid_value_name, False, {'value_restrict'})
+])
+def test_restrict_value(key, subkey, arch, value_name, triggers_event, tags_to_apply,
+                  get_configuration, configure_environment, restart_syscheckd,
+                  wait_for_fim_start):
+    """
+    Check the only files detected are those matching the restrict regex
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_registry/test_restrict/test_registry_restrict.py:61]: This test is about registry values, not files. Please, changes this comment.
```suggestion
    Check the only registry values detected are those matching the restrict regex.
```

File: tests/integration/test_fim/test_registry/test_restrict/test_registry_restrict.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    (key, sub_key_2, valid_subkey, KEY_WOW64_64KEY, True, {'key_restrict'}),
+    (key, sub_key_1, no_valid_subkey, KEY_WOW64_64KEY, False, {'key_restrict'}),
+    (key, sub_key_2, no_valid_subkey, KEY_WOW64_32KEY, False, {'key_restrict'}),
+    (key, sub_key_2, no_valid_subkey, KEY_WOW64_64KEY, False, {'key_restrict'})
+    ])
+def test_restrict_key(key, subkey, test_subkey, arch, triggers_event, tags_to_apply,
+                 get_configuration, configure_environment, restart_syscheckd,
+                  wait_for_fim_start):
+    """
+    Check the only files detected are those matching the restrict regex
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_registry/test_restrict/test_registry_restrict.py:145]: This test is about registry keys, not files. Please, change it.
```suggestion
    Check the only registry keys detected are those matching the restrict regex.
```

**Review by jesusjimsa — CHANGES_REQUESTED** [review]

**Review by jesusjimsa — CHANGES_REQUESTED** [review]
One of the cases failed in my machine.
```
================================================= test session starts =================================================
platform win32 -- Python 3.7.3, pytest-5.1.2, py-1.8.0, pluggy-0.13.0
rootdir: C:\Users\jmv74211\Desktop\wazuh-qa\tests\integration, inifile: pytest.ini
plugins: html-2.0.1, metadata-1.10.0, testinfra-5.0.0
collected 24 items

test_restrict\test_registry_restrict.py ......ssssssssssss.F....                                                 [100%]

====================================================== FAILURES =======================================================
___ test_restrict_key[get_configuration1-HKEY_LOCAL_MACHINE-SOFTWARE\\testkey-restrict_key-512-True-tags_to_apply1] ___

key = 'HKEY_LOCAL_MACHINE', subkey = 'SOFTWARE\\testkey', test_subkey = 'restrict_key', arch = 512
triggers_event = True, tags_to_apply = {'key_restrict'}
get_configuration = {'apply_to_modules': ['test_registry_restrict'], 'metadata': {'fim_mode': 'scheduled', 'windows_reg
stry_1': 'HKEY_LOC...ibutes': [...], 'value': 'HKEY_LOCAL_MACHINE\\SOFTWARE\\testkey'}}], 'section': 'syscheck'}], 'tag
': ['key_restrict']}
configure_environment = None, restart_syscheckd = None, wait_for_fim_start = None

    @pytest.mark.parametrize('key, subkey, test_subkey, arch, triggers_event, tags_to_apply', [
        (key, sub_key_1, valid_subkey, KEY_WOW64_64KEY, True, {'key_restrict'}),
        (key, sub_key_2, valid_subkey, KEY_WOW64_32KEY, True, {'key_restrict'}),
        (key, sub_key_2, valid_subkey, KEY_WOW64_64KEY, True, {'key_restrict'}),
        (key, sub_key_1, no_valid_subkey, KEY_WOW64_64KEY, False, {'key_restrict'}),
        (key, sub_key_2, no_valid_subkey, KEY_WOW64_32KEY, False, {'key_restrict'}),
        (key, sub_key_2, no_valid_subkey, KEY_WOW64_64KEY, False, {'key_restrict'})
        ])
    def test_restrict_key(key, subkey, test_subkey, arch, triggers_event, tags_to_apply,
                     get_configuration, configure_environment, restart_syscheckd,
                      wait_for_fim_start):
        """
        Check the only files detected are those matching the restrict regex

        Parameters
        ----------
        key : str
            Root key (HKEY_*)
        subkey : str
            Path of the registry.
        test_subkey : str
            Name of the key that will be used for the test
        arch : str
            Architecture of the registry.
        triggers_event : bool
            True if an event must be generated, False otherwise.
        tags_to_apply : set
            Run test if match with a configuration identifier, skip otherwise.
        """
        check_apply_test(tags_to_apply, get_configuration['tags'])
        test_key = os.path.join(subkey, test_subkey)
        create_registry(registry_parser[key], test_key, arch)

        # Go ahead in time to let syscheck perform a new scan
        check_time_travel(get_configuration['metadata']['fim_mode'] == 'scheduled', monitor=wazuh_log_monitor)

        if triggers_event:
            event = wazuh_log_monitor.start(timeout=global_parameters.default_timeout,
                                callback=callback_detect_event, accum_results=1).result()
            assert event['data']['type'] == 'added', f'Event type not equal'
            assert event['data']['path'] == os.path.join(key, test_key), f'Event path not equal'
            assert event['data']['arch'] == '[x32]' if arch == KEY_WOW64_32KEY else '[x64]', f'Arch not equal'

        else:
            while True:
                ignored_key = wazuh_log_monitor.start(timeout=global_parameters.default_timeout,
                                                       callback=callback_restricted,
                                                       error_message='Did not receive expected '
                                                                     '"Sending FIM event: ..." event').result()
                if ignored_key == os.path.join(key, subkey):
                    break
        delete_registry(registry_parser[key], test_key, arch)
        # Go ahead in time to let syscheck perform a new scan
        check_time_travel(get_configuration['metadata']['fim_mode'] == 'scheduled', monitor=wazuh_log_monitor)

        if triggers_event:
            event = wazuh_log_monitor.start(timeout=global_parameters.default_timeout,
>                                       callback=callback_detect_event, accum_results=1).result()

test_restrict\test_registry_restrict.py:190:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\jmv74211\AppData\Local\Programs\Python\Python37-32\lib\site-packages\wazuh_testing-4.0.0-py3.7.egg\wazuh_testi
g\tools\monitoring.py:189: in start
    error_message=error_message).result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <wazuh_testing.tools.monitoring.QueueMonitor object at 0x041661B0>, timeout = 10
callback = <function callback_detect_event at 0x041193D8>, accum_results = 1, update_position = True, timeout_extra = 0
error_message = ''

    def start(self, timeout=-1, callback=_callback_default, accum_results=1, update_position=True, timeout_extra=0,
              error_message=''):
        """Start the queue monitoring until the stop method is called."""
        if not self._continue:
            self._continue = True
            self._abort = False
            result = None

            while self._continue:
                if self._abort:
                    self.stop()
                    if error_message:
                        logger.error(error_message)
                        logger.error(f"Results accumulated: "
                                     f"{len(result) if isinstance(result, list) else 0}")
                        logger.error(f"Results expected: {accum_results}")
>                   raise TimeoutError()
E                   TimeoutError

C:\Users\jmv74211\AppData\Local\Programs\Python\Python37-32\lib\site-packages\wazuh_testing-4.0.0-py3.7.egg\wazuh_testi
g\tools\monitoring.py:461: TimeoutError
------------------------------------------------ Captured stderr call -------------------------------------------------
2020-10-29 13:26:12,515 - wazuh_testing - INFO - Creating registry key HKEY_LOCAL_MACHINE\SOFTWARE\testkey\restrict_key
2020-10-30 02:26:12,000 - wazuh_testing - INFO - Changing the system clock from 2020-10-29 13:26:12.515625 to 2020-10-3
 02:26:12
2020-10-30 02:26:13,000 - wazuh_testing - INFO - Removing registry key [x32]HKEY_LOCAL_MACHINE\SOFTWARE\testkey\restric
_key
2020-10-30 15:26:13,000 - wazuh_testing - INFO - Changing the system clock from 2020-10-30 02:26:13 to 2020-10-30 15:26
13
-------------------------------------------------- Captured log call --------------------------------------------------
INFO     wazuh_testing:fim.py:409 Creating registry key HKEY_LOCAL_MACHINE\SOFTWARE\testkey\restrict_key
INFO     wazuh_testing:fim.py:1329 Changing the system clock from 2020-10-29 13:26:12.515625 to 2020-10-30 02:26:12
INFO     wazuh_testing:fim.py:556 Removing registry key [x32]HKEY_LOCAL_MACHINE\SOFTWARE\testkey\restrict_key
INFO     wazuh_testing:fim.py:1329 Changing the system clock from 2020-10-30 02:26:13 to 2020-10-30 15:26:13
====================================== 1 failed, 11 passed, 12 skipped in 55.25s ======================================
```

**Review by jesusjimsa — CHANGES_REQUESTED** [review]
The test keeps failing to me.
[test_registry.zip](https://github.com/wazuh/wazuh-qa/files/5452568/test_registry.zip)

**Review by jesusjimsa — APPROVED** [review]
LGTM!

**Review by bah07 — APPROVED** [review]
LGTM!

**Pull request merged**